Fix tiebreak when loading blocks from disk (and add tests for comparing chain ties)

sr-gi opened the pull request:
This PR grabs some interesting bits from https://github.com/bitcoin/bitcoin/pull/29284 and fixes some edge cases in how block tiebreaks are dealt with.

## Regarding #29284

The main functionality from the PR was dropped given it was not an issue anymore, however, reviewers pointed out some comments were outdated https://github.com/bitcoin/bitcoin/pull/29284#discussion_r1522023578 (which to my understanding may have led to thinking that there was still an issue) it also added test coverage for the aforementioned case which was already passing on master and is useful to keep.

## New functionality

While reviewing the superseded PR, it was noticed that blocks that are loaded from disk may face a similar issue (check https://github.com/bitcoin/bitcoin/pull/29284#issuecomment-1994317785 for more context).

The issue comes from how tiebreaks for equal work blocks are handled: if two blocks have the same amount of work, the one that is activatable first wins, that is, the one for which we have all its data (and all of its ancestors'). The variable that keeps track of this, within `CBlockIndex` is `nSequenceId`, which is not persisted over restarts. This means that when a node is restarted, all blocks loaded from disk are defaulted the same `nSequenceId`: 0. 
Now, when trying to decide what chain is best on loading blocks from disk, the previous tiebreaker rule is not decisive anymore, so the `CBlockIndexWorkComparator` has to default to its last rule: whatever block is loaded first (has a smaller memory address). 

This means that if multiple same work tip candidates were available before restarting the node, it could be the case that the selected chain tip after restarting does not match the one before.

Therefore, the way `nSequenceId` is initialized is changed to:

- 0 for blocks that belong to the previously known best chain
- 1 to all other blocks loaded from disk

sipa committed:
test: add functional test for complex reorgs

sr-gi committed:
Updates CBlockIndexWorkComparator outdated comment

mzumsande commented on src/validation.cpp:
```diff
@@ -4189,7 +4189,7 @@ bool ChainstateManager::ProcessNewBlockHeaders(const std::vector<CBlockHeader>&
             const CBlockIndex& last_accepted{**ppindex};
             const int64_t blocks_left{(GetTime() - last_accepted.GetBlockTime()) / GetConsensus().nPowTargetSpacing};
-            const double progress{100.0 * last_accepted.nHeight / (last_accepted.nHeight + blocks_left)};
+            const double progress{100.0 * (last_accepted.nHeight + 1) / (last_accepted.nHeight + 1 + blocks_left)};
```
If `blocks_left` can be negative, it doesn't look like the fix really solves the issue. `blocks_left` might as well be `-2` instead of `-1`, and we divide by zero again. Might be better to check if it's zero and, if so, skip the log?

sr-gi replied:
I'll remove it from here and rebase on top of that once merged

sr-gi committed:
Set the same best tip on restart if two candidates have the same work

Before this, if we had two (or more) same work tip candidates and restarted our node,
it could be the case that the block set as tip after bootstrap didn't match the one
before stopping. That's because the work and `nSequenceId` of both block will be the same
(the latter is only kept in memory), so the active chain after restart would have depended
on what tip candidate was loaded first.

This makes sure that we are consistent over reboots.

sr-gi committed:
test: Adds block tiebreak over restarts tests

Adds tests to make sure we are consistent on activating the same chain over
a node restart if two or more candidates have the same work when the node is shutdown

sr-gi commented:
Rebased to drop the custom log fix in favor of a more generic solution (https://github.com/bitcoin/bitcoin/pull/29640#discussion_r1523327586)

luke-jr commented on src/node/blockstorage.cpp:
```diff
@@ -163,7 +163,7 @@ bool CBlockIndexWorkComparator::operator()(const CBlockIndex* pa, const CBlockIn
 
     // Use pointer address as tie breaker (should only happen with blocks
-    // loaded from disk, as those all have id 0).
+    // loaded from disk, as those all have id 1).
```
0 or 1

luke-jr replied:
Actually, I guess the 0s should be unique so this just needs rephrasing some other way

sr-gi replied:
Updated to:

(should only happen with blocks loaded from disk, as those share the same nSequenceId: 0 for blocks on the best chain, 1 for all others)

sr-gi commented on src/validation.cpp:
```diff
@@ -4501,6 +4501,22 @@ bool Chainstate::LoadChainTip()
+    // Otherwise the scoring will be based on the memory address location instrad of the nSequenceId
+    setBlockIndexCandidates.erase(tip);
+    TryAddBlockIndexCandidate(tip);
+    // We also need to re-prune blocks that now score potentially worse than before
```
Umm, I really cannot see why I wanted to do this in two steps. It really doesn't seem necessary.

I've also re-arranged the code a bit to minimize the diff

sr-gi commented:
Rebased to deal with CI failing.

c76f4c6dab267b8f83c75e44939f77d237251c1a has been amended to include `__file__` in the `ChainTiebreaksTest` constructor, as required by any subclass of `BitcoinTestFramework` since https://github.com/bitcoin/bitcoin/pull/30463

mzumsande reviewed (commented):
Code Review ACK c6ca2a17ea4d76f21d7702f70ba892a9494576bf
I agree that it is more consistent to keep the same tip over a restart in case of multiple candidates (although I don't think this is much of a problem in practice).

sr-gi commented on test/functional/feature_chain_tiebreaks.py:
```diff
@@ -0,0 +1,103 @@
+        assert_equal(node.getbestblockhash(), blocks[2].hash)
+
+        self.log.info('Send parents B3-B4 of B8-B10 in reverse order')
+        peer.send_blocks_and_test([blocks[4]], node, success=False, force_send=True)
```
I guess it may be worth crating an issue / opening a separate PR to fix this, given it is unrelated

TheCharlatan replied:
It would be an easy doc fix, something like:
```diff
-         - if success is True: assert that the node's tip advances to the most recent block
-         - if success is False: assert that the node's tip doesn't advance
+         - if success is True: assert that the node's tip is the last block in blocks at the end of the operation.
+         - if success is False: assert that the node's tip isn't the last block in blocks at the end of the operation
```
Will submit separately if it does not get picked up here.

sr-gi replied:
I took this. [3cb689a](https://github.com/bitcoin/bitcoin/pull/29640/commits/3cb689adf58e3b415775a57dbb79fd1d0cabab80)

achow101 commented on test/functional/feature_chain_tiebreaks.py:
```diff
@@ -99,5 +99,53 @@ def run_test(self):
+        peer.send_blocks_and_test([blocks[1]], node, success=False)
+
+        self.log.info('Restart the node and check that the best tip before restarting matched the ones afterwards')
+        # Restart and check enough times to this to eventually fail if the logic is broken
+        for _ in range(10):
+            self.restart_node(0)
+            assert_equal(blocks[0].hash, node.getbestblockhash())
```
In c6ca2a17ea4d76f21d7702f70ba892a9494576bf "test: Adds block tiebreak over restarts tests"

This does not seem to fail on master, but it does on this branch with the fix commit reverted. Is it possible that this was fixed by a different PR?

mzumsande replied:
for me, it also fails on master - but not always, intermittently.

furszy commented on src/chain.h:
```diff
@@ -192,7 +192,9 @@ class CBlockIndex
 
     //! (memory only) Sequential id assigned to distinguish order in which blocks are received.
-    int32_t nSequenceId{0};
+    //! Initialized to 1 when loading blocks from disk, except for blocks belonging to the best chain
+    //! which overwrite it to 0.
+    int32_t nSequenceId{1};
```
In 6756af191f0:
nit: It might be more maintainable to define two constants, such as `ID_DISK-ONLY_BEST_CHAIN` and `ID_DISK-ONLY_FORK` (or something similar).

furszy reviewed (commented):
Concept ACK.

Just started reviewing. It needs to be rebased on master to include CMake support.
Also, left a small nit along the way.

sr-gi commented:
Thanks @furszy. I've added two constants for the values loaded from disk and rebased over master.

furszy commented:
> I've added two constants for the values loaded from disk

Missed to push them?

sr-gi committed:
Make nSequenceId init value constants

Make it easier to follow what the values come without having to go
over the comments, plus easier to maintain

sr-gi commented:
> > I've added two constants for the values loaded from disk
> 
> Missed to push them?

Indeed. Should be fixed now

sr-gi commented:
The last commit had a small typo, force-pushing to fix it

furszy commented on src/validation.cpp:
```diff
@@ -4718,9 +4718,22 @@ bool Chainstate::LoadChainTip()
     m_chain.SetTip(*pindex);
+    tip = m_chain.Tip();
+
+    // Make sure our chain tip before shutting down scores better than any other candidate
+    // to maintain a consistent best tip over reboots
```
```suggestion
    // Make sure our chain tip before shutting down scores better than any other candidate
    // to maintain a consistent best tip over reboots in case of a tie with another chain.
```

furszy commented on src/validation.cpp:
```diff
@@ -4718,9 +4718,22 @@ bool Chainstate::LoadChainTip()
+    while (target) {
+        target->nSequenceId = 0;
+        target = target->pprev;
+    }
```
q: cannot just set the tip's sequence id to 0?

We only compare the tip inside `TryAddBlockIndexCandidate` and not the entire chain (if not, then a test might be missing because setting only the tip to 0 still passes all tests).

mzumsande replied:
I don't think that this is an edge case we'd necessarily need to support though - it was just the only case I could come up where it would matter.

mzumsande replied:
I still think it would be sufficient to only set it for the tip - I don't think that the edge case of `invalidateblock` justifies doing another iteration over all blocks during each init (it takes 0.1 - 0.2s on mainnet for me on a reasonably fast laptop, which is not that much, but I would imagine it might be quite a bit longer with slower hardware / longer chains such as testnet3 - unless there is another reason to set it for all blocks that I am unaware of.

sr-gi replied:
IIRC, the rationale here was that `nSequenceId` is supposed to be monotonically increasing, so having a chain of blocks where the previous blocks all have nSequenceId = `x-1` and the tip has `x` was weird. Since this is something that should only happen on restarts, the extra fraction of a second was not that relevant.

I'm happy to change it so only the tip has 0 if you think the time savings are more relevant than the inconsistency with the design. Comments will also need to be updated.

sr-gi replied:
> so having a chain of blocks where the previous blocks all have nSequenceId = x-1 and the tip has x

Sorry, I wrote that wrong.  `:s/x-1/x+1/g`

In the case were we only change the tip, the rest of blocks will have `nSequenceId=1`, so we will end up with a chain like:

`1 -> 1 -> 1 -> 0 (tip)` that will evolve to `1 -> 1 -> 1 -> 0 -> 2 (tip)`

All other chains will have `nSequenceId=1` from start to end.

In the case proposed here, we'll have 

`0 -> 0 -> 0 -> 0 (tip)` evolving to `0 -> 0 -> 0 -> 0 -> 2 (tip)`

mzumsande replied:
oh right, I got confused in my earlier post, you are right. Still, maybe others could chime in if they think that this monotonicity is worth the added work or not.

furszy reviewed (commented):
Code review ACK 177d07f

ryanofsky reviewed (commented):
Might be ready for merge if @mzumsande and @sipa can re-ack

mzumsande commented on test/functional/feature_chain_tiebreaks.py:
```diff
@@ -0,0 +1,103 @@
+        node.invalidateblock(blocks[9].hash)
+        node.invalidateblock(blocks[10].hash)
+        # B7 is now active.
+        assert_equal(node.getbestblockhash(), blocks[7].hash)
```
Is this assertion correct - why can't the best blockhash be B8?
Even though B7 is received first, both are put into `std::multimap<CBlockIndex*, CBlockIndex*> m_blocks_unlinked;` (with no comparator, so pointer address is used which is decided by the OS?!). and later, when their parent arrives, accessed via `equal_range` in `ReceivedBlockTransactions` where `nSequenceId` is then set in that order. Couldn't this be done in the reverse order B8 -> B7 if the OS gives out the pointer addresses in a different order?

TheCharlatan replied:
I'm not sure I am following here. How does `ReceivedBlockTransactions` come into play here? As far as I can tell this activates the best chain, which should then correctly pick B7 again through `FindMostWorkChain`.

sr-gi replied:
I completely overlooked this, but it looks like (unhinted) insertion order is preserved when retrieving via `equal_range`:

> Since emplace and unhinted insert always insert at the upper bound, the order of equivalent elements in the equal range is the order of insertion unless hinted insert or emplace_hint was used to insert an element at a different position.

source: https://en.cppreference.com/w/cpp/container/multimap/equal_range

mzumsande replied:
Ah, I missed this - then it should be fine!

sr-gi committed:
test: Fixes send_blocks_and_test docs

It's not true that if success=False the tip doesn't advance. It doesn'test
advance to the provided tip, but it can advance to a competing one

sr-gi commented on src/chain.h:
```diff
@@ -35,6 +35,9 @@ static constexpr int64_t MAX_FUTURE_BLOCK_TIME = 2 * 60 * 60;
  * MAX_FUTURE_BLOCK_TIME.
  */
 static constexpr int64_t TIMESTAMP_WINDOW = MAX_FUTURE_BLOCK_TIME;
+ //! Init values for CBlockIndex nSequenceId when loaded from disk
```
Covered in [3869469](https://github.com/bitcoin/bitcoin/pull/29640/commits/38694694201ca3c374d4a17f3764ea617a70ddfe)

sr-gi commented:
Thanks for reviewing @TheCharlatan and @furszy. I rebased the code and addressed the outstanding comments.

sr-gi commented on test/functional/feature_chain_tiebreaks.py:
```diff
@@ -0,0 +1,151 @@
+            blocks[-1].solve()
+
+        # Send blocks and test the last one is not connected
+        self.log.info('Send A1 and A2. Make sure than only the former connects')
```
Fixed in [8d73371](https://github.com/bitcoin/bitcoin/pull/29640/commits/8d7337189a1b7024a40933bf082b7b574fd007e1)

sr-gi commented:
I think all feedback has been addressed @mzumsande @sipa @furszy @TheCharlatan @maflcko 

Please let me know if there's anything else pending or if anything doesn't make sense.

sr-gi commented on test/functional/feature_chain_tiebreaks.py:
```diff
@@ -0,0 +1,151 @@
+        # Construct B1-B10.
+        for i in range(1, 11):
+            blocks.append(create_block(
+                hashprev=int(blocks[(i - 1) >> 1].hash, 16),
```
Rebased and fixed

sipa reviewed (commented):
utACK 0465574c127907df9b764055a585e8281bae8d1d

Happy with either the "whole active chain is set to nSequence 0" or the "only tip is set to nSequence 0" approaches.

sedited reviewed (approved):
ACK 0465574c127907df9b764055a585e8281bae8d1d

sr-gi commented:
Would you mind giving this another look @mzumsande @furszy ?

furszy reviewed (commented):
Tested ACK 0465574c127907df9b764055a585e8281bae8d1d.

Reverting the fix makes the introduced test fails (after ~60 test runs).
